File: aider_is_programmable_2.js
import { spawn, execSync } from 'child_process';
import { mkdir, existsSync } from 'fs';
import { dirname } from 'path';

// Create directory if it doesn't exist
const todoDir = './cc_todo';
const todoFile = `${todoDir}/todo.ts`;

try {
  // Ensure directory exists
  if (!existsSync(todoDir)) {
    mkdir(todoDir, { recursive: true }, (err) => {
      if (err) throw err;
    });
  }

  // Generate branch name
  const branchName = 'feature-todo-app';

  // 1. Create and checkout a new branch
  console.log(`Creating and checking out new branch: ${branchName}`);
  execSync(`git checkout -b ${branchName}`);

  // 2. Run aider directly with the todo task
  console.log('Running aider to create todo app...');
  const aiderProcess = spawn('aider', [
    '--no-git',
    todoFile,
    '--message',
    'CREATE ./cc_todo/todo.ts: a zero library CLI todo app with basic CRUD.'
  ], {
    stdio: 'inherit'
  });

  // Handle aider completion
  aiderProcess.on('close', (code) => {
    if (code !== 0) {
      console.error(`Aider process exited with code ${code}`);
      cleanupAndExit(1);
      return;
    }

    try {
      // 3. Git operations - stage and commit
      console.log('Staging and committing changes...');
      execSync('git add .');
      execSync('git commit -m "Add TypeScript todo app with CRUD functionality"');

      // 4. Switch back to main branch
      console.log('Switching back to main branch...');
      execSync('git checkout main');

      console.log(`Task completed. Changes committed to branch: ${branchName}`);
    } catch (error) {
      console.error(`Error during git operations: ${error.message}`);
      cleanupAndExit(1);
    }
  });

} catch (error) {
  console.error(`Error: ${error.message}`);
  cleanupAndExit(1);
}

// Helper function to cleanup and exit
function cleanupAndExit(code) {
  try {
    execSync('git checkout main');
  } catch (error) {
    console.error('Failed to switch back to main branch');
  }
  process.exit(code);
}

File: aider_is_programmable_2.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.9"
# dependencies = []
# ///

import subprocess
import os

# Create directory if it doesn't exist
todo_dir = "./cc_todo"
os.makedirs(todo_dir, exist_ok=True)
todo_file = f"{todo_dir}/todo.ts"

# Generate a random branch name
branch_name = f"feature-todo-app"

try:
    # 1. Create and checkout a new branch
    print(f"Creating and checking out new branch: {branch_name}")
    subprocess.run(["git", "checkout", "-b", branch_name], check=True)

    # 2. Run aider directly with the todo task
    print("Running aider to create todo app...")
    aider_cmd = [
        "aider",
        "--no-git",  # We'll handle git ourselves
        todo_file,
        "--message",
        "CREATE ./cc_todo/todo.ts: a zero library CLI todo app with basic CRUD.",
    ]
    subprocess.run(aider_cmd, check=True)

    # 3. Git operations - stage and commit
    print("Staging and committing changes...")
    subprocess.run(["git", "add", "."], check=True)
    subprocess.run(
        ["git", "commit", "-m", "Add TypeScript todo app with CRUD functionality"],
        check=True,
    )

    # 4. Switch back to main branch
    print("Switching back to main branch...")
    subprocess.run(["git", "checkout", "main"], check=True)

    print(f"Task completed. Changes committed to branch: {branch_name}")

except subprocess.CalledProcessError as e:
    print(f"Command failed: {e}")
except Exception as e:
    print(f"Error: {e}")
    # Try to return to main branch if something went wrong
    try:
        subprocess.run(["git", "checkout", "main"], check=True)
    except:
        pass
